torch/csrc/jit/mobile/backport_manager.cpp

```diff
@@ -0,0 +1,238 @@
+// a function can be reused by other backport functions, move it to the utility
+// function group. It will be easier to split out backport_manager.cpp to
+// smaller files when it grows too long.
+
+// The functions needed for backport model from v5 to v4.
+namespace {
+
+void writeArchiveV4(
+    PyTorchStreamWriter& writer,
+    const std::string& archive_name,
+    const c10::IValue& value) {
+  std::vector<char> data;
+
+  // Vector to capture the run-time class types during pickling the IValues
+  std::vector<c10::ClassTypePtr> memoizedClassTypes;
+  Pickler data_pickle(
+      [&](const char* buf, size_t size) {
+        data.insert(data.end(), buf, buf + size);
+      },
+      nullptr,
+      nullptr,
+      &memoizedClassTypes);
+  data_pickle.protocol();
+  data_pickle.pushIValue(value);
+  data_pickle.stop();
+  size_t i = 0;
+  std::string prefix = archive_name + "/";
+
+  for (const auto& td : data_pickle.tensorData()) {
+    WriteableTensorData writable_td = getWriteableTensorData(td);
+    std::string fname = prefix + c10::to_string(i++);
+    writer.writeRecord(fname, writable_td.data(), writable_td.sizeInBytes());
+  }
+  std::string fname = archive_name + ".pkl";
+  writer.writeRecord(fname, data.data(), data.size());
+}
+
+bool backport_v5_to_v4(
+    PyTorchStreamReader& reader,
+    PyTorchStreamWriter& writer) {
+  // 1) read from archive `bytecode` archive
+  std::vector<IValue> bytecode_values = get_bytecode_values(reader);
+  if (!check_bytecode_version(bytecode_values, kBytecodeVersionV5)) {
+    TORCH_WARN("Incorrect bytecode version for input model.");
+    return false;
+  }
+
+  // 2) Copy everything except bytecode related to new output
+  copy_non_bytecode(reader, writer);
+
+  // 3) write `bytecode` archive
+  // Update the bytecode version in bytecode.pkl
+  update_bytecode_version(bytecode_values, kBytecodeVersionV4);
+  // Construct the list of ivalues to a big tuple
+  auto bytecode_tuple = c10::ivalue::Tuple::create(std::move(bytecode_values));
+  // write `bytecode` archive
+  writeArchiveV4(writer, "bytecode", bytecode_tuple);
+  return true;
+}
+} // namespace
+
+// A generic contract for backport logic to the previous bytecode version.
+// Args:
+// * PyTorchStreamReader has access to the input model from N bytecode version.
+// * PyTorchStreamWriter has access to the output model backported to the
+// previous N-1 bytecode version. Returns true if successful, false otherwise.
+using BytecodeBackportFunction = std::function<bool(
+    caffe2::serialize::PyTorchStreamReader&,
+    caffe2::serialize::PyTorchStreamWriter&)>;
+
+BackportManager::BackportManager() {
+  registerBytecodeBackportFunction(kBytecodeVersionV5, backport_v5_to_v4);
+}
+
+std::unordered_map<
+    int64_t,
+    std::function<bool(
+        caffe2::serialize::PyTorchStreamReader&,
+        caffe2::serialize::PyTorchStreamWriter&)>>&
+BackportManager::bytecodeBackportFunctions() const {
```

Comment:
Can this static instance be member of BackportManager? Then you just have static instance of backport manager.
Also does it need to be thread safe? It should be ok but maybe we should make it thread safe. Or make sure that there are no race conditions.

Reply:
Any good suggestion to make it thread safe? Also would like to know which specific part do you think might run into race conditions.

Reply:
Actually I take that back. I think we should be fine. Only place where race condition may happen is when registration is happening from two different callsites. Each modifying the map. Another one while doing registration and then doing backport at the same time. But since registration happens inside the constructor and thats the only place through which it can happen, so the first one should be good. Second one should be fine as well since, backport cannot happen before backport_manager is constructed. So this is a non issue. Forget all that I said.

Reply:
Thanks for clear explanation. Good to learn!